// @tencent/txv-history-manage/lib/request.js
var consts = require('./consts')
var jsonp = require('@tencent/jsonp')
var fns = require('@tencent/fns')

var MAX_PN = 10       //分页拿取一次最多拿10条

//list的可选组合：cid_vid或者_或者_和pn,cur的组合,cid_vid|cid_vid
function getHistory(key, cb, options) {
  var data, url;
  if (key.length > 2) {
    data = {
      list: key
    }
    url = consts.api.specifyview
  } else {
    data = {    //通过cgi去全量拉取数据,或者提供翻页和翻页条数
      pn: 0,
      cur: 0
    }
    url = consts.api.batchview     //cgi的URL
    if (options.pn && options.pn > MAX_PN) {
      options.pn = MAX_PN
    }
    data = fns.extend(data, options)
  }
  jsonp(url, data, function (err, res) {
    if (err) {
      cb(err)
    } else if (res && res.result && res.result.code < 0) {
      cb('请求失败')
    } else {
      res = res && res.list || []
      res.length > 0 ? cb(null, res) :
        cb('不存在' + key + '的历史记录')
    }
  })
}
//cid_vid|cid_vid
// function batchHistory(keyidArr, cb) {
//     var list = keyidArr.join('|')
//     var data = {
//         simp: 1,
//         pver: 1,
//         plat: 2,
//         t: consts.rParamT.BATCHREQUESTINFO,
//         list: list
//     }
//     jsonp(URL, data, function(err, res) {
//         if (err) {
//             cb(err)
//         } else if (res && res.result && res.result.code < 0) {
//             cb('请求失败')
//         } else {
//             res = res && res.list || []
//             res.length > 0 ? cb(null, res) :
//                 cb('不存在' + list + "的历史记录")
//         }
//     })
// }

module.exports = {
  getHistory: getHistory
}
