// @tencent/txmv-core/share.js
var ua = require("@tencent/txv.ua")();

var jsbridge = require('./jsbridge');

var str_shareUrl = "";
var str_shareImg = "";

var str_timeLineTitle = "";
var str_friendTitle = "";
var str_friendSubtitle = "";

var _callback = [];

var exp = {
	//如此定义参数位置，是为了兼容之前的接口，保证在base.min未更新的时候，仍然不会出现问题
	init: function (i_str_shareUrl, i_str_shareImg, i_str_timeLineTitle, i_str_friendSubtitle, i_str_friendTitle) {
		jsbridge(always);

		function always() {
			exp.initParam(i_str_shareUrl, i_str_shareImg, i_str_timeLineTitle, i_str_friendSubtitle, i_str_friendTitle);
			if (ua.browser.WeChat) {
				exp.shareToWx();
			} else if (ua.browser.MQQClient) {
				exp.shareToQQ();
			} else if (ua.browser.QQvideo) {
				exp.shareToApp();
			} else if (ua.browser.qqnews) {
				exp.shareToNews();
			}

			exp.init = function (i_str_shareUrl, i_str_shareImg, i_str_timeLineTitle, i_str_friendSubtitle, i_str_friendTitle) {
				exp.initParam(i_str_shareUrl, i_str_shareImg, i_str_timeLineTitle, i_str_friendSubtitle, i_str_friendTitle);
				if (ua.browser.MQQClient) {
					exp.shareToQQ();
				} else if (ua.browser.QQvideo) {
					exp.shareToApp();
				} else if (ua.browser.qqnews) {
					exp.shareToNews();
				}
			}
		}
	},
	registerCallback: function (fn) {
		typeof fn == 'function' && _callback.push(fn);
	},
	initParam: function (i_str_shareUrl, i_str_shareImg, i_str_timeLineTitle, i_str_friendSubtitle, i_str_friendTitle) {
		str_shareUrl = i_str_shareUrl || location.href;
		str_shareImg = i_str_shareImg || null;

		str_timeLineTitle = i_str_timeLineTitle || "腾讯视频";
		str_friendTitle = i_str_friendTitle || "腾讯视频";
		str_friendSubtitle = i_str_friendSubtitle || "";
	},
	shareToWx: function () {
		if (window.WeixinJSBridge) {
			exp.handleInWxShare();
		}
		else {
			document.addEventListener('WeixinJSBridgeReady', function () {
				exp.handleInWxShare();
			});
		}
	},
	handleInWxShare: function () {
		this.useWxNewShare();
		this.useWxOldShare();
	},
	useWxNewShare: function () {
		WeixinJSBridge.on("menu:general:share", function (argv) {
			var content = "#" + str_friendTitle + "#" + str_friendSubtitle;
			var str_shareTitle = "";
			var str_shareDesc = "";
			if (!!ua.os.ios) {
				content += str_shareUrl;
			}
			if (argv.shareTo == "friend" || argv.shareTo == 'QQ') {
				str_shareTitle = str_friendTitle;
				str_shareDesc = str_friendSubtitle;
			} else {
				str_shareTitle = str_timeLineTitle;
			}
			argv.generalShare({
				//"type": "video",
				"content": content,
				"title": str_shareTitle,
				"desc": str_shareDesc,
				"img_url": str_shareImg,
				"img_width": "120",
				"img_height": "90",
				"link": str_shareUrl,
				"data_url": str_shareUrl,
				appid: "wxca942bbff22e0e51"
			}, function (res) {
				_callback.forEach(function (fn) {
					fn(res);
				})
			});
		});
	},
	useWxOldShare: function () {
		window.shareConfig = {
			img_url: str_shareImg,
			img_width: 120,
			img_heigth: 90,
			link: str_shareUrl,
			title: str_friendTitle,
			desc: str_friendSubtitle
		};
		WeixinJSBridge.on("menu:share:timeline", function (e) {
			var str_shareTitle = str_timeLineTitle;
			var str_shareDesc = "";
			var data = {
				appid: "wxca942bbff22e0e51",
				img_url: str_shareImg,
				img_width: "120",
				img_height: "90",
				link: str_shareUrl,
				desc: str_shareDesc,
				title: str_shareTitle
			};
			WeixinJSBridge.invoke("shareTimeline", data, function (res) {
				_callback.forEach(function (fn) {
					fn(res);
				})
			});
		});
		WeixinJSBridge.on("menu:share:appmessage", function (e) {
			var str_shareTitle = str_friendTitle;
			var str_shareDesc = str_friendSubtitle;
			WeixinJSBridge.invoke("sendAppMessage", {
				appid: "wxca942bbff22e0e51",
				img_url: str_shareImg,
				img_width: "120",
				img_height: "90",
				link: str_shareUrl,
				desc: str_shareDesc,
				title: str_shareTitle
			}, function (res) {
				_callback.forEach(function (fn) {
					fn(res);
				})
			});
		});
		WeixinJSBridge.on("menu:share:weibo", function () {
			var content = "#" + str_friendTitle + "#" + str_friendSubtitle;
			if (!!ua.os.ios) {
				content += str_shareUrl;
			}

			WeixinJSBridge.invoke("shareWeibo", {
				"content": content,
				"url": str_shareUrl
			}, function (res) {
				_callback.forEach(function (fn) {
					fn(res);
				})
			});
		});
	},
	shareToQQ: function () {
		if (window.mqq && window.mqq.data && window.mqq.data.setShareInfo) {
			var info = {
				name: str_friendTitle || "腾讯视频",
				description: str_friendSubtitle || " ",
				image: str_shareImg
			};
			window.mqq.data.setShareInfo({
				share_url: str_shareUrl,
				title: info.name,
				desc: info.description,
				image_url: str_shareImg
			}, function (res) {
				setTimeout(function () {
					console.log(res);
				}, 3000)
			});
			if (window.Zepto) {
				var uahead = Zepto('head');
				Object.keys(info).forEach(function (key) {
					var uaelem = uahead.find('meta[itemprop="' + key + '"]');
					if (uaelem.length) {
						uaelem.attr('content', info[key]);
					} else {
						Zepto('<meta itemprop="' + key + '" content="' + info[key] + '"/>').appendTo(uahead);
					}
				});
			}
		}
	},
	shareToApp: function () {
		TenvideoJSBridge.invoke('setMoreInfo', {
			hasShare: true,
			hasRefresh: true,
			shareInfo: {
				title: str_friendTitle,
				subTitle: str_friendSubtitle,
				singleTitle: str_timeLineTitle,
				url: str_shareUrl,
				imageUrl: str_shareImg
			}
		})
	},
	shareToNews: function () {
		if (window.TencentNews && window.TencentNews.setShareArticleInfo) {
			window.TencentNews.setShareArticleInfo(
				str_friendTitle,
				str_timeLineTitle,
				str_friendSubtitle,
				str_shareUrl,
				str_shareImg
			);
		}
	}
};
module.exports = exp;
