// @tencent/txmv-core/monitor/boss-tvfe.js
'use strict';
var urlParam = require('../util/urlParam');
var cookie = require('../util/cookie');
var report = require('./report');
var lazyreport = require('./lazyreport');
var noop = function () {}
var uuidGetter = noop
//http://beehive.boss.webdev.com/bossid/myid/viewId?id=4795#bossid
/**
 * 前端组自建，类tcss上报
 */

//有引用jq或zepto的话自动绑定全局事件
var $;
try {
    $ = window.jQuery || window.Zepto || (window.tvp && tvp.$);
} catch(e) {}
if ($) {
    $(document.body).on('click', '[_boss]', function (e) {
        var $target = $(e.currentTarget);
        hot($target.attr('_boss'));
    });
}

//获取默认ptag
var sessionStorage = window.sessionStorage || {getItem: function() {return ''}};
var ptag = urlParam.get('ptag') || sessionStorage.getItem('ptag') || cookie.get('ptag');
var openid = urlParam.get('openid') || cookie.get('openid');

var baseUrl = '//btrace.qq.com/kvcollect?BossId=3255&Pwd=2118551257&s_url=' +
    encodeURIComponent(location.href.split('?')[0]) + '&s_referrer=' + encodeURIComponent(document.referrer) +
    '&s_ua=' + navigator.userAgent +
    '&s_uin=' + (cookie.get('o_cookie') || cookie.get('uin') || cookie.get('luin')) +
    '&s_openid=' + openid;

function getBaseUrl(){
    return baseUrl + '&s_uuid=' + (uuidGetter() || UID()) + '&_dc=' + Math.random()
}

// report(b\aseUrl + '&s_ptag=' + ptag);

/**
 * 生成一个UUID
 */
function guid() {
  function s4() {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  }
  return [s4(), s4(), s4(), s4()].join('')
}
/**
 * @from txmv.base.mypv
 * 生成或从cookie获取uuid
 */
var _uid
function UID() {
    if (_uid) return _uid

    var uid = cookie.get('tvfe_boss_uuid');
    if (!uid) {
        uid = guid()
        try {
            cookie.set('tvfe_boss_uuid', uid, 'qq.com', false, 87600)
        } catch(e) {
            cookie.set('tvfe_boss_uuid', uid, '', false, 87600)
        }
    }
    _uid = uid
    return uid
}

function hot(name, cfg) {
    name = name.split('.');
    if (name.length < 3 || name.length > 5) {
        console.warn('按钮名分段数不对，可能会导致数据丢失');
    }
    var url = getBaseUrl();
    url += '&s_ptag=' + ptag +
        '&s_app=' + (name[0] || '') +
        '&s_module=' + (name[1] || '') +
        '&s_behave=' + (name[2] || '') +
        '&s_sub=' + (name[3] || '') +
        '&s_fifth=' + (name[4] || '');
    if (cfg && cfg.lazy) {
        lazyreport.push(url);
    } else {
        report(url);
    }
}


module.exports = {
    uuid: function () {
        return UID()       
    },
    uuidGetter: function (fn) {
        uuidGetter = fn || noop 
    },
    //自行设置ptag
    ptag: function(str) {
        ptag = str;
    },
    hot: hot,
    uid: UID
};
