// @tencent/txmv-core/monitor/tcss.js
/**
 * 点击流系统
 */
var getScript = require("../util/getScript");
var Lazyreport = require("./lazyreport");

var jspath = ('https:' == location.protocol ? "https://pingjs.qq.com/ping_video.js" : "http://pingjs.qq.com/ping_video.js");

var getTCSS = (function () {
	var ready = false;
	var callbacks = [];
	getScript(
		jspath,
		function () {
			ready = true;
			var cb;
			while (cb = callbacks.shift()) {
				cb();
			}
		}
	);
	return function (cb) {
		if (ready) {
			cb();
		} else {
			callbacks.push(cb);
		}
	}
})();

/**
 * 初始化PGV点击流系统上报的一些全局变量
 */
function initPGVGlobalParam() {
	window.pgvVirtualDomain = "";
	window.pvCurDomain = "";
	window.pvCurUrl = "";
}

var button = function (key, virtualPage, virtualDomain, lazy) {
	if (typeof key != "string") return;

	if (!virtualPage) {
		var pagename = key;
		var arrKey = key.split(".");
		if (arrKey.length > 0) {
			pagename = arrKey[0];
		}
		pagename += ".html";
		virtualPage = "/virtualpage/" + pagename;
	}
	initPGVGlobalParam();
	var sendData = {
		hottag: 'txmv.' + key,
		virtualURL: virtualPage
	};

	if (virtualDomain) {
		sendData['virtualDomain'] = virtualDomain;
	}
	if (lazy) {
		Lazyreport.push("pgvSendClick(JSON.parse('" + JSON.stringify(sendData) + "'))");
	} else {
		getTCSS(function () {
			window.pgvSendClick && pgvSendClick(sendData);
		});
	}
};

/**
 * 检测 _hot 的链路上的a标签是否满足 Lazyreport 如果链路上没有a标签，则在当前页面上报 如果链路上有a标签，如果是
 * 新开页面、站外链接、没有链接 ，则在当前页面直接上报，否则 Live.Lazyreport.push(item)
 *
 */
var checkHotLazyreport = function (ele, hot) {
	if (!ele || !hot)
		return;
	var href = ele.getAttribute('href'),
		target = ele.getAttribute('target');
	href = !!href ? href : '';
	target = !!target ? target : '';
	/*_blank,_self,_parent,_top,框架名，这几个，忽略框架名，就只有_blank不会让当前页面跳转*/
	if ('_blank' == target || !href || href.indexOf('javascript') >= 0 || (href.indexOf('http://') == 0 && href.indexOf(location.hostname) < 0)) {
		button(hot);
	} else {
		button(hot, false, false, true);
	}
};

/**
 * 绑定热点统计事件 查找用用_hotkey类名的元素，获取_hot属性作为热点统计值
 */
var bindHotkey = function (zepto) {
	zepto(document).off('click', '[_hot]').on('click', '[_hot]', function () {
		checkHotLazyreport(this, this.getAttribute('_hot'));
	}, true);
};

module.exports = {
	pv: function () {
		getTCSS(function () {
			window.pgvMain && pgvMain();

			//上报上一个页面延时上报的数据
			Lazyreport.report();
		});
	},
	button: button,
	checkHotLazyreport: checkHotLazyreport,
	bindHotkey: bindHotkey
};
