// @tencent/txmv-core/login/controller/qq.js
/**
 * 手Q以及其它浏览器
 */
'use strict';
var cookie = require('../../util/cookie');
var jsonp = require("../../util/jsonp");
var time33 = require("../../helper/time33");

exports.getLoginData = function (opt, cb) {
	if (typeof opt == 'function' && !cb) {
		cb = opt;
		opt = null;
	}
	var data = {
		skey: cookie.get('skey'),
		uin: cookie.get('uin'),
		lskey: cookie.get('lskey'),
		luin: cookie.get('luin')
	};

	// 强制推到下一个eventloop执行
	setTimeout(function () {
		cb(null, data);
	});
	if (data.skey || data.lskey) {
		data.main_login = 'qq';

		data._skey = data.skey || data.lskey;
		data._uin = parseInt((data.uin || data.luin).replace(/^o0*/g, ""), 10);
	}
};
exports.getLoginDataSync = exports.getLoginData.sync = function () {
	var data = {
		skey: cookie.get('skey'),
		uin: cookie.get('uin'),
		lskey: cookie.get('lskey'),
		luin: cookie.get('luin')
	};
	if (data.skey || data.lskey) {
		data.main_login = 'qq';

		data._skey = data.skey || data.lskey;
		data._uin = parseInt((data.uin || data.luin).replace(/^o0*/g, ""), 10);
	}
	return data;
};

exports.gTk = function (cb) {
	this.getLoginData(function (err, data) {
		var tokens = {};
		if (data.skey || data.lskey) {
			tokens.g_tk = time33(data.skey || data.lskey);
			tokens.query = 'g_tk=' + tokens.g_tk;
		}

		cb(null, tokens);
	})
};

var cacheUserInfo = null;

exports.getUserInfo = function (cb) {
	this.getLoginData(function (err, data) {
		var lowlogin = 0;
		if (!data.skey && !data.lskey) {
			return cb('not login');
		} else if (!data.skey) {
			lowlogin = 1;
		}

		if (cacheUserInfo) {
			return cb(null, cacheUserInfo);
		}
		jsonp('//video.qq.com/fcgi-bin/get_userinfo?' +
			'otype=json&' +
			'_=' + Date.now() + '&' +
			'type=2&' +
			'low_login=' + lowlogin,
			function (err, data) {
				if (!err && data.ret == 0 && data.data[0]) {
					cacheUserInfo = {
						face: data.data[0].headurl,
						nick: data.data[0].nick
					};
					cb(null, cacheUserInfo);
				} else {
					cb(err || data);
				}
			});
	});
};


var default_conf = {
	link_target: 'blank',
	low_login: 0,
	target: 'self',
	style: 9,
	appid: '532001601',
	s_url: window.location.href
};
exports.openLogin = function (cfg) {
	cfg = cfg || {};
	var url = '//ui.ptlogin2.qq.com/cgi-bin/login?';
	
	for(var k in default_conf) {
		if(default_conf.hasOwnProperty(k)) {
			typeof cfg[k] === 'undefined' ? cfg[k] = default_conf[k] : void(0);
		}
	}

	cfg.s_url && (cfg.s_url = encodeURIComponent(cfg.s_url));

	var params = [];
	for (var key in cfg) {
		if (cfg.hasOwnProperty(key)) {
			params.push(key + '=' + cfg[key])
		}
	}

	url += params.join('&');
	setTimeout(function () {
		window.location.href = url;
	}, 200);
};
